From 6f07e5f06c8712acc423485f657799fc8e11e56c Mon Sep 17 00:00:00 2001
From: Xin Long <lucien.xin@gmail.com>
Date: Sun, 31 Mar 2019 22:50:08 +0800
Subject: [PATCH] tipc: check bearer name with right length in tipc_nl_compat_bearer_enable
Git-commit: 6f07e5f06c8712acc423485f657799fc8e11e56c
Patch-mainline: v5.1-rc4
References: bsc#1051510

Syzbot reported the following crash:

Bug: KMSAN: uninit-value in memchr+0xce/0x110 lib/string.c:961  memchr+0xce/0x110 lib/string.c:961  string_is_valid net/tipc/netlink_compat.c:176 [inline]  tipc_nl_compat_bearer_enable+0x2c4/0x910 net/tipc/netlink_compat.c:401  __tipc_nl_compat_doit net/tipc/netlink_compat.c:321 [inline]  tipc_nl_compat_doit+0x3aa/0xaf0 net/tipc/netlink_compat.c:354  tipc_nl_compat_handle net/tipc/netlink_compat.c:1162 [inline]  tipc_nl_compat_recv+0x1ae7/0x2750 net/tipc/netlink_compat.c:1265  genl_family_rcv_msg net/netlink/genetlink.c:601 [inline]  genl_rcv_msg+0x185f/0x1a60 net/netlink/genetlink.c:626  netlink_rcv_skb+0x431/0x620 net/netlink/af_netlink.c:2477  genl_rcv+0x63/0x80 net/netlink/genetlink.c:637  netlink_unicast_kernel net/netlink/af_netlink.c:1310 [inline]  netlink_unicast+0xf3e/0x1020 net/netlink/af_netlink.c:1336  netlink_sendmsg+0x127f/0x1300 net/netlink/af_netlink.c:1917  sock_sendmsg_nosec net/socket.c:622 [inline]  sock_sendmsg net/socket.c:632 [inline]

Uninit was created at:
  __alloc_skb+0x309/0xa20 net/core/skbuff.c:208
  alloc_skb include/linux/skbuff.h:1012 [inline]
  netlink_alloc_large_skb net/netlink/af_netlink.c:1182 [inline]
  netlink_sendmsg+0xb82/0x1300 net/netlink/af_netlink.c:1892
  sock_sendmsg_nosec net/socket.c:622 [inline]
  sock_sendmsg net/socket.c:632 [inline]

It was triggered when the bearer name size < TIPC_MAX_BEARER_NAME,
it would check with a wrong len/TLV_GET_DATA_LEN(msg->req), which
also includes priority and disc_domain length.

This patch is to fix it by checking it with a right length:
'TLV_GET_DATA_LEN(msg->req) - offsetof(struct tipc_bearer_config, name)'.

Reported-by: syzbot+8b707430713eb46e1e45@syzkaller.appspotmail.com
Signed-off-by: Xin Long <lucien.xin@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 net/tipc/netlink_compat.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/net/tipc/netlink_compat.c b/net/tipc/netlink_compat.c
index 4ad3586da8f0..5f8e53cca222 100644
--- a/net/tipc/netlink_compat.c
+++ b/net/tipc/netlink_compat.c
@@ -397,7 +397,12 @@ static int tipc_nl_compat_bearer_enable(struct tipc_nl_compat_cmd_doit *cmd,
 	if (!bearer)
 		return -EMSGSIZE;
 
-	len = min_t(int, TLV_GET_DATA_LEN(msg->req), TIPC_MAX_BEARER_NAME);
+	len = TLV_GET_DATA_LEN(msg->req);
+	len -= offsetof(struct tipc_bearer_config, name);
+	if (len <= 0)
+		return -EINVAL;
+
+	len = min_t(int, len, TIPC_MAX_BEARER_NAME);
 	if (!string_is_valid(b->name, len))
 		return -EINVAL;
 
-- 
2.16.4

